Feature/scheduled doc snapshots

ejfrancis opened the pull request:
for #12 
- created a ProseMeteorServer class that instantiates all server code
- added snapshotIntervalMs to ProseMeteorServer constructor (not sure if it'll stay here in the long run, but it demonstrates that it can be changed and passed down to Authority instances)
- changed Authority to create an interval to write doc snapshots to db if changes have occurred since the last snapshot was written

The big change here is that all document steps are now persisted to db, so you can shut down the server and restart it and the document will still have its state. Previously it was only stored in memory. You'll see a message logged to server console for now that says when the write happened, so you can see when/if it writes. It should only happen if the interval is hit AND if changes have occurred since last snapshot.

@funkyeah

ejfrancis committed:
moved main server code into api/prosemeteor/server ProseMeteorServer class

ejfrancis committed:
setup authority to write snapshots to db on provided snapshotIntervalMs, added stopInterval method (not used yet)

ejfrancis commented on lib/imports/startup/server/server-index.js:
not sure if we should ask the integrator of our package to create this in their own code and pass configs, i.e.

`const proseMeteorServer = new ProseMeteorServer({ snapshotIntervalMs: 1000 });`

or if we should create the server instance internally, and allow the integrator to pass in config on startup or at runtime with

`ProseMeteor.setSnapshotInterval(1000)`

if we allow them to change it whenever they want at runtime, it may be a bit tricker than if we only allow them to set it at startup

gckellerman commented on lib/imports/api/authority/server/authority.js:
```diff
@@ -88,25 +96,72 @@ export class Authority {
+    }
+    return version;
+  }
+
```
why wouldn't this always be this.steps.length or `latestDocState()`?

```
  latestDocState() {
    return {
      version: this.steps.length,
      docJSON: this.doc.toJSON()
    };
  }
```

ejfrancis replied:
Since the doc can be loaded from a snapshot, just using `this.steps.length` doesn't work anymore since our previous implementation assumed all steps are in memory. If I open a doc and it loads it's last snapshot, which was at version 1200, and then I make 5 edits, `this.steps.length=5` but the actual doc version is 1205 (the latest snapshot version 1200 + the 5 edits since then). I actually overlooked that part in `latestDocState`, I'll have to change it to use `this.version()`

ejfrancis replied:
I think I'll change `numberOfSnapshots()` to not query db for number of snapshots each time as well since that's a lot of reads. We can just load the number of snapshots from db when a doc is opened and store it in memory, then just increment the number in memory every time we call `storeSnapshot`.

gckellerman commented on lib/imports/startup/server/server-index.js:
I like the idea of having them pass config and have control when things start.

gckellerman commented on lib/imports/api/authority/server/authority.js:
```diff
@@ -88,25 +96,72 @@ export class Authority {
+  /**
+  * Starts the timer interval to store doc snapshots to db.
+  */
+  startSnapshotInteval() {
```
typo in the func name and call above

ejfrancis committed:
added CollabStep and CollabStepClientId classes to store a historical collab step and client id, so that we can search through them and find those for a specific version. changed Authority version and stepsSince logic to search through and find those according to the version they were applied to. added arguments checks to Authority and AuthorityManager

ejfrancis committed:
change ProseMeteorServer instantiation to be done by the integrator, in demo/server/main.js

ejfrancis commented:
alright @funkyeah thanks for the initial code review, I just pushed up some more changes
- changed it so the integrator has to create a `new ProseMeteorServer()`, so that options can be passed to constructor (now it's in `demo/server/main.js`)
- fixed `snapshotInterval()` name typo
- changed Authority `stepsSince()` and `version` logic to always be correct, regardless of whether or not a snapshot was loaded or how many steps are currently in memory
  - created a `CollabStep` class that represents a step on a collaborative document (stores the version it was applied to)
  - similarly, created `CollabStepClientId` class
  - changed `stepsSince()` to search through all CollabSteps/CollabStepClientIds to find those that were after version. it no longer relies on all steps being in memory since the beginning of a doc
- added argument checks to `Authority` and `AuthorityManager` via Meteor's "check" package (already helped me catch some issues)

These are changes around the core collab logic, not very complicated but still changing the core logic. Feel free to test it out and see how it works for you

ejfrancis commented on lib/imports/startup/server/server-index.js:
k just added that in my latest commits

ejfrancis commented on lib/imports/api/authority/server/authority.js:
```diff
@@ -83,30 +103,84 @@ export class Authority {
+    let steps = [];
+    let clientIds = [];
+    // iterate over all steps and find the steps/stepClientIds applied since specified version
+    for (let i = 0; i < this.steps.length; i++) {
```
note that this will take increasingly longer time as more steps are in memory, i.e. O(N) complexity. we'll definitely need to limit the number of steps held in memory to counter it. there may be some optimizations/shortcuts possible here to not always have to iterate over all steps

gckellerman commented on lib/server/server-main.js:
```diff
@@ -2,3 +2,7 @@
+
+import { ProseMeteorServer as ProseMeteorServerClass } from './../imports/api/prosemeteor/server/prosemeteor-server';
+
+export let ProseMeteorServer = ProseMeteorServerClass;
```
@ejfrancis You can replace lines 6 and 8 with just the following instead of doing the rename dance. That's probably my fault since I did the same thing in the original repo. I think its clearer though to export directly.

`export { ProseMeteorServer } from './../imports/api/prosemeteor/server/prosemeteor-server';`

ejfrancis replied:
thanks, I was looking around a ton of es2015 docs to try to find the correct syntax for that and couldn't find it anywhere

gckellerman commented on lib/imports/api/prosemeteor/server/prosemeteor-server.js:
```diff
@@ -0,0 +1,49 @@
+  /*
+  * Set up Meteor methods for communication with clients.
+  */
+  setupClientMethods() {
```
can this be moved into the authority-manager where the other client methods are setup?

ejfrancis replied:
yup making that change in my next commit, will be in soon

ejfrancis committed:
move openDoc method to AuthorityManager, fix ProseMeteorServer export

funkyeah commented:
I'm definitely cool with merging this.
